### ISSUE TITLE
Loss of API status on Forbidden

### ISSUE THREAD

**Issue by enj** [openshift/origin#16731]
Now:

```
$ curl -k https://$(openshift start --print-ip):8443/api/v1/namespaces/myproject/pods/podX1:8080/proxy/appEndPoint
User "system:anonymous" cannot get pods/unsafeproxy in the namespace "myproject": User "system:anonymous" cannot get pods/unsafeproxy in project "myproject"
```

In the past:

```
$ curl -k https://$(openshift start --print-ip):8443/api/v1/namespaces/myproject/pods/podX1:8080/proxy/appEndPoint
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "User \"system:anonymous\" cannot get pods/unsafeproxy in project \"myproject\"",
  "reason": "Forbidden",
  "details": {
    "name": "podX1:8080",
    "kind": "pods"
  },
  "code": 403
}
```

Are we OK with losing our nice machine parseable error?  We lost this by moving to the upstream handlers.  We should probably fix upstream to do something more useful.

@openshift/sig-security

**Comment by smarterclayton** [comment]
I can't see any reason why we shouldn't fix upstream.

**Comment by liggitt** [comment]
similar issue reported upstream, I verified we get full status objects back in 1.8, hadn't tracked down where it was fixed

https://github.com/kubernetes/kubernetes/issues/52754#issuecomment-331682105

**Comment by surajssd** [comment]
This seems to be fixed in upstream:

on k8s
```
$ curl -k https://localhost:6443/api/v1/namespaces/default/pods/
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "pods is forbidden: User \"system:anonymous\" cannot list pods in the namespace \"default\"",
  "reason": "Forbidden",
  "details": {
    "kind": "pods"
  },
  "code": 403
}
```

on openshift

```
# curl -k https://localhost:8443/api/v1/namespaces/myproject/pods/
User "system:anonymous" cannot list pods in the namespace "myproject": User "system:anonymous" cannot list pods in project "myproject"
```

by upstream do you mean Kubernetes or something else?

**Comment by liggitt** [comment]
yes, kubernetes. fixed in the 1.8 timeframe, which means openshift would pick up the fix in 3.8

**Comment by surajssd** [comment]
Is this a beginner friendly issue? If yes, what particular areas I should look at to solve this problem?

I have been trying to solve this issue for quiet sometime now! 

- First I tried to find where this string is built? But it seems like the code is very modular and simple grep won't help.
- When I tried the same on k8s, I found that the server throws the error in a [rbac.go](https://github.com/openshift/origin/blob/c7d2994db4a4dbe0ef3ea42e05c6a79d6fe5fd08/vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/rbac.go#L116), about the denial of the request
- Later I have inferred that I am being authenticated as `system:anonymous` but authorization step fails (please correct me if I am wrong here), the request is not going to the **Admission Controllers**
- Doing more grepping on the source, found this one particular function [`Forbidden`](https://github.com/openshift/origin/blob/048cccc32c7f38f622834864600e562c540aa0a3/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/errors.go#L44), thought this might be called but it is not being called so, also found a similar function [here](https://github.com/openshift/origin/blob/ecaeef8665f16e230ec069e25a64963b57ce3f00/pkg/cmd/server/handlers/authorization.go#L39) which is also not being called.
- Now I resorted to see what kubectl has to offer, so ran the command `kubectl get nodes --v 10`, which listed bunch of functions, but not sure if those are server side or only called client side. The output has following line:

```
I1016 18:08:20.723527    8790 round_trippers.go:436] GET https://localhost:8443/api/v1/nodes 403 Forbidden in 0 milliseconds
```
which comes from [here](https://github.com/openshift/origin/blob/048cccc32c7f38f622834864600e562c540aa0a3/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/client-go/transport/round_trippers.go#L436)

- This also led me nowhere?

Is there some specific place I should look at to fix this issue?

**Comment by surajssd** [comment]
@liggitt thanks for that reply, did not see your early comment, but I think this will help me understand how this was solved :+1:  :-)

**Comment by enj** [comment]
This was fixed in https://github.com/kubernetes/kubernetes/pull/47384

@smarterclayton @liggitt are we OK with 3.7 not having status errors on forbidden?  3.6- work as expected, and so will 3.8+  Seems strange to have 3.7 be the only release with this in a broken state.

**Comment by xiaods** [comment]
any update?

**Comment by simo5** [comment]
Seem like it will be automatically fixed in 3.8, I am going to close this issue as we are not going to deal with this in 3.7 timeframe.

**Comment by smarterclayton** [comment]
So we are going to regress our public API?

**Comment by eparis** [comment]
yes

**Comment by smarterclayton** [comment]
Please make sure this is documented in the release notes for API changes and that it will be resolved in 3.8
